File: restaurants/taj-mahal/get_menu.py
from urllib import request
from bs4 import BeautifulSoup
import json
import datetime

from os import path
directory = path.dirname(path.realpath(__file__))


def main():
    html = request.urlopen("https://www.tajmahalgbg.se/lunch-meny/").read().decode()

    html_tree = BeautifulSoup(html, features="html.parser")

    relevant_html = html_tree.find_all(class_="elementor-widget-wrap")[0].find_all(["h4", "p"])


    now = datetime.datetime.now().isocalendar()

    menu = {
        "year": now[0],
        "week": now[1],
        "days": {}
    }

    for day in range(0, 5):
        i = day * 12

        dishes = []
        
        for dish_index in range(5):
            try:
                title = relevant_html[i + dish_index * 2].string.strip()
                description = relevant_html[i + dish_index * 2 + 1].string.strip()
                dish = {
                    "title": title,
                    "description": description
                }

                dishes.append(dish)


            except IndexError:
                pass
        
        if len(dishes) == 0:
            dishes = ["N/A"]

        menu["days"][day] = dishes
    
    fp = open(directory + "/menu.json", "w", encoding="utf-8")

    json.dump(menu, fp, ensure_ascii=False)

    
main()

File: restaurants/taj-mahal/getMenu.js
const cheerio = require("cheerio")
const https = require("https")

function getHtml() {
    let options = {
        hostname: 'www.tajmahalgbg.se',
        port: 443,
        path: "/lunch-meny/",
        method: "GET"
    }
    
    return new Promise((resolve, reject) => {
        let req = https.request(options, (res) => {
            let data = ""
            res.on("data", d => data += d)
            res.on("end", () => {
                resolve(data)
            })
        })
        
        req.end()
    })
}




module.exports = async () => {
    let html = await getHtml()
    const $ = cheerio.load(html)


    let relevantHtml = $(".elementor-widget-wrap h4, .elementor-widget-wrap p")

    let now = new Date()

    let menu = {
        restaurant: "Taj Mahal",
        year: now.getFullYear(),
        week: getWeek(now),
        days: {}
    }

    
    for (let day = 0; day < 5; day++) {
        let i = day * 12
        let dishes = []
        for (let dishIndex = 0; dishIndex < 5; dishIndex++) {
            try {
                let title = relevantHtml[i + dishIndex * 2].children[0].data.trim()
                let description = relevantHtml[i + dishIndex * 2 + 1].children[0].data.trim()

                dishes.push({
                    title: title,
                    description: description
                })
            } catch (e) {
                // Do nothing
            }
        }
        menu.days[day] = dishes
    }

    return menu
}


function getWeek(date) {
    var year = date.getFullYear();
    var oneJan = new Date("Jan 1 " + year);
    var oneJanDay = (oneJan.getDay() + 6) % 6 
    var day = Math.floor((date.getTime() - oneJan.getTime()) / 86400000);
    var week = Math.floor(day / 7);

    if (oneJanDay < 4) {
        week++;
    }
    return week;
}
